File: merge-sort.clj
(ns merge-sort)

(defn merge-sort 
  "Sorts a sequence with merge sort by the given predicate. If none is 
given, <= is used. O(n log n) performance."
  ([coll] (merge-sort coll <=))
  ([coll cmp]
  (letfn [(first-half [xs] (take (/ (count xs) 2) xs))
          (second-half [xs] (drop (/ (count xs) 2) xs))
          (merge [xs ys] (cond (not (seq xs)) ys
                               (not (seq ys)) xs
                               (cmp (first xs) (first ys)) (cons (first xs) (merge (rest xs) ys))
                               :else (cons (first ys) (merge xs (rest ys)))))]
    (if (>= 1 (count coll)) coll
        (merge (merge-sort (first-half coll) cmp) (merge-sort (second-half coll) cmp))))))

File: mergesort.hs
module MergeSort (mergeSort) where

merge :: (Ord a) => [a] -> [a] -> [a] -> [a]
merge [] [] xs = xs
merge ls [] xs = xs ++ ls
merge [] rs xs = xs ++ rs
merge (l:ls) (r:rs) xs = if l < r 
                         then merge ls (r:rs) (xs ++ [l]) 
                         else merge (l:ls) rs (xs ++ [r])

mergeSort :: (Ord a) => [a] -> [a]
mergeSort xs 
          | length xs < 2 = xs
          | otherwise = merge ls rs []
                      where n  = div (length xs) 2
                            ls = mergeSort $ take n xs
                            rs = mergeSort $ drop n xs